### DISCUSSION TITLE
Symbols Indexing: Local vs Global

### DISCUSSION THREAD

**Discussion by tajmone** [tajmone/Sublime-PML#14] [General]
@pml-lang and @celtic-coder, I wanted to share with you some thoughts on Symbols indexing, and get your feedback on the issue.

Currently, Sublime PML is indexing all symbols only locally (i.e. for the specific file in which they occur). Sublime Text allows indexing a Symbol either locally or globally (for the whole Project), or both.

The idea is that in multi-file sources, e.g. a C/C++ project, you might want be able to jump to a symbol definition (e.g. a function) even if the definition is in another file within the Project. On the other hand, when dealing with documents you usually want to index title headings only locally to the file in which they are found.

With PML it's hard to make a clean cut decision, because the `[insert` node allows to include external PML docs into each other, a feature which is probably much used in large document project (e.g. splitting a document by chapters).

I think the question is whether to index symbols only locally or both locally and globally (local indexing seems to be always necessary, I can't think of any valid exceptions rights now).

On the user side, Sublime Text allows to quickly navigate through either [local symbols] or [global symbols].

The Symbols panel will show all indexed symbols, not just PML's symbols (e.g. markdown symbols, including headings, and tokens from the various programming languages inside the project). For this reason, adding any Symbol to the Global Index can quickly lead to index pollution, although filtering through the desired results should be quite easy, especially since Sublime PML adorns PML symbols by manipulating their appearance.

In my opinion, Chapter titles should only be indexed locally, even though some projects do split chapters across different files it probably makes little sense to index them globally. But I might be wrong on this, for I haven't yet worked on big PML documentation projects — so it's an open issue for now.

I'm also inclined to believe that parameters (occurrences and definitions alike) should be indexed both locally and globally, for I foresee that defining their values in a commonly shared file is going to be a common strategy, so it's going to be useful to be able to jump to their definition, even if in an external file (or more than one, in case of conditional inclusion of one or the other).

In any case, it's worth noting that ST always allows to lookup just the local Symbols index (<kbd>Ctrl</kbd><kbd>R</kbd>), so global indexing doesn't prevent local lookup operations.

I'd like to ear your thoughts and ideas on this, and what you envision would be the optimal editing experience when it comes to deciding which symbols (nodes, attributes) are worth indexing and how, and which potential features might be built around them.

[local symbols]: https://docs.sublimetext.io/guide/usage/file-management/navigation.html#goto-anything-operators "Learn more at Sublime Text Unofficial Documentation"
[global symbols]: https://docs.sublimetext.io/guide/usage/file-management/navigation.html#goto-symbol-in-project "Learn more at Sublime Text Unofficial Documentation"

**Comment by pml-lang** [comment]
Whenever I write an article in PML I use several files in a single directory or a directory tree, where sub-directories represent parent chapters. Having small files makes it easier to maintain them and move chapters around. See [example](https://github.com/ppl-lang/blog/tree/main/2021-03-Suggestion_for_better_XML_HTML/PML_source) of a directory tree.

Therefore I would expect symbols to be indexed globally.

> C/C++ project, you might want be able to jump to a symbol definition (e.g. a function) even if the definition is in another file within the Project.

Yes, and a analogous case in PML would be this: A parameter is defined in file A, and I want to use it in file B. Or: File C contains an internal link (xref) to a chapter in file D.

> parameters (occurrences and definitions alike) should be indexed both locally and globally, for I foresee that defining their values in a commonly shared file is going to be a common strategy

I agree.

Sometimes people will of course use just one file for an article (e.g. small article), and then indexing globally might be overkill. As so often, there is no one-size-fits-all solution. But _most_ users would probably prefer global indexing.

> deciding which symbols (nodes, attributes) are worth indexing

That's a good question. I don't have enough experience myself to confidently answer this, but I would say that at a minimum, all parameters and all nodes that have an explicit id (e.g. `[code id=hello_example]`) should be indexed.

**Reply by celtic-coder** [reply]
Hi Christian (@pml-lang), Tristano (@tajmone),

I am experimenting with using the [Leo outlining editor](https://leoeditor.com/) to create both the outline and the content of a document on which I am now working. Since I am able to move around the nodes in the outline, I don't need the facility to move files around in the file system to modify the document, as the final output will be fairly short. 

However, for notes that I took in a college course from a few years back, it would have been better to have split up the AsciiDoctor document into several standalone files. Now that I am using PML, and when I have a requirement for a larger final output, having the facility to lookup symbols in a PML project would be an absolute requirement.

Kind Regards,
Liam

**Reply by tajmone** [reply]
@pml-lang:

> Sometimes people will of course use just one file for an article (e.g. small article), and then indexing globally might be overkill.

In that case the user would simply lookup for local symbols using <kbd>Ctrl</kbd><kbd>R</kbd> (instead of <kbd>Ctrl</kbd>Shift<kbd>R</kbd>). I wouldn't worry about the overkill, Sublime Text has been used with huge code projects including thousands of code lines and hundred of files, and overall it does a great job of tracking symbols in the Index.

Sometimes indexing can become slow, if a project is handling thousands of symbols; and if you change the [project indexing settings]  (e.g. exclude/include patterns) it might become necessary to either restart Sublime Text (for Index changes to take effect) or to destroy the old Index database, so it can be rebuild from scratch. These issues aside, I guess that most PML projects should cope fine with indexing; as for big code projects where PML is used just for the documentation aspect, the general considerations of any ST project apply (i.e. PML not bearing any particular weight on the usual workflow).

> As so often, there is no one-size-fits-all solution. But most users would probably prefer global indexing.

I agree with that, although there might be some exception for particular nodes and attributes which might require local indexing only — e.g. Chapter titles; otherwise you might have multiple "Introduction" or "Preface" headings showing up globally, since most publications will have that; but maybe this is exactly what some users might be looking for (i.e. a way to quickly find all same-named sections in a project), which is why I'd like to hear about how different users would normally use these features in daily work.

But then again, ST also shows you the file path when there are two identical symbols, and if they share the same file name it will show their full path, so ST is quite friendly when dealing with same-named symbols. So, maybe the real question is if global indexing could, after all, potentially pollute the index and stand in the way of optimal editing, or whether ST interfaces are smart enough so that would not even be a problem in practical use.

> I would say that at a minimum, all parameters and all nodes that have an explicit id (e.g. `[code id=hello_example]`) should be indexed.

I totally agree on this: all identifiers should be indexed to provide quick access.



[project indexing settings]: https://www.sublimetext.com/docs/3/projects.html "Sublime Text official documentation"

**Reply by tajmone** [reply]
@celtic-coder:

> I am experimenting with using the Leo outlining editor to create both the outline and the content of a document on which I am now working.

Leo is a wonderful tool. I came across it in the past and was impressed by it, it's so much more than just and editor; but ultimately didn't adopt it due to lack of time to learn to use it properly.

> Since I am able to move around the nodes in the outline, I don't need the facility to move files around in the file system to modify the document, as the final output will be fairly short.

I guess that's an advantage of using outlining tools over classical editors.

Which reminds of [frame-based editing], yet another approach to editing source files and documents, which probably would work well with the PML syntax. There aren't many frame editors available ([Greenfoot], [BlueJ]), but that's an interesting paradigm when it comes to working with structured documents using lightweight markup syntaxes.

> Now that I am using PML, and when I have a requirement for a larger final output, having the facility to lookup symbols in a PML project would be an absolute requirement.

Good, so far we all feel the same about the benefits of global indexing.

The final question then is _what_ to index. So, feel free to mention any nodes and attributes which you think are worth indexing, and why.

So far, [Sublime PML indexes the following symbols]:

- Parameters
- Chapter titles
- Identifiers

Also, you might have noticed that it manipulates the symbols before indexing them, which I thought makes them more recognizable and thus provides a better editing experience — but I didn't get any feedback on this aspect of the syntax, and whether users like the idea at all. Other syntaxes do similar cosmetic changes (e.g. the markdown package also adds `#`s to indexed headings).



[Sublime PML indexes the following symbols]: https://github.com/tajmone/Sublime-PML#symbols-indexing "Read the details"

[frame-based editing]: https://www.greenfoot.org/frames/
[Greenfoot]: https://www.greenfoot.org/download
[BlueJ]: https://bluej.org/

**Reply by pml-lang** [reply]
> So far, Sublime PML indexes the following symbols:
    Parameters
    Chapter titles
    Identifiers

That's a good start. Besides adding nodes with an `id` attribute, maybe you could also add nodes that have a `title` attribute (but no id) defined in the text (for nodes like `table`, `table_data`, `code`, `input`, `output`, etc.)

> manipulates the symbols before indexing them, which I thought makes them more recognizable

Good idea. I never used symbols in ST on my old PC, but could they be prefixed with the node's and/or attribute's name, so that they are sorted in the list of symbols?

**Reply by tajmone** [reply]
> you could also add nodes that have a title attribute (but no id) defined in the text (for nodes like `table`, `table_data`, `code`, `input`, `output`, etc.)

Yes, that would be useful, especially in big documents split into modules, for it would allow to quickly find and __Goto__ a node by typing its title.

In this case, it would also make sense to decorate the title Symbols according to the node they belong to, e.g. by adding some prefix (either the node `[nodename ` or a symbol commonly associated with that context) for this would allow to quickly filter the results based on the node type, which quickly and drastically reduces the results lists, and Fuzzy matching only requires typing a few chars to pinpoint any result.

> I never used symbols in ST on my old PC, but could they be prefixed with the node's and/or attribute's name, so that they are sorted in the list of symbols?

You can manipulate Symbols at indexing time by passing them through [one or more RegEx substitutions] using [Oniguruma] syntax, which is a powerful RegEx engine. So, yes, we have good control over Symbols transformations. 

Here's a screenshot of the current __Goto Symbol__ panel with Sublime PML:

![Goto Symbol Screenshot](https://user-images.githubusercontent.com/11064292/116080974-fa087f00-a699-11eb-81e4-f3a03b2432a9.png)




[one or more RegEx substitutions]: https://docs.sublimetext.io/reference/symbols.html#text-transformations "Sublime Text unofficial documentation"

[Oniguruma]: https://github.com/kkos/oniguruma/blob/master/doc/RE

**Reply by tajmone** [reply]
As you might notice from the above screenshot, quoted titles are indexed without the surrounding quotes, and any escaped quote is also converted to a simple quote.

Of course, it would be nice if we could expand parameters in index strings and contents, but the current ST3 API doesn't expose this functionality (but I believe ST4 will). In any case, this would not always be possible, e.g. because a parameter might be defined conditionally, or via CLI options, etc., so this might not be so useful in Symbol Indexing, but it would have been nice to offer a pop-up with the expanded parameter when hovering with the mouse on a `<<param>>` in the source.

**Reply by pml-lang** [reply]
The screenshot looks good and promising! I like the colors too.

> not always be possible, e.g. because a parameter might be defined conditionally, or via CLI options

True!
